here-naksha-lib-ext-manager/src/main/java/com/here/naksha/lib/extmanager/ExtensionCache.java

```diff
@@ -112,41 +112,70 @@ private void publishIntoCache(KVPair<Extension, File> result, ExtensionConfig ex
       if (!isNullOrEmpty(extension.getInitClassName())) {
         try {
           Class<?> clz = loader.loadClass(extension.getInitClassName());
-          clz.getConstructor(INaksha.class, Extension.class).newInstance(naksha, extension);
-        } catch (ClassNotFoundException
-            | InvocationTargetException
-            | InstantiationException
-            | NoSuchMethodException
-            | IllegalAccessException e) {
+          Object obj = clz.getConstructor(INaksha.class, Extension.class).newInstance(naksha, extension);
+          if (obj instanceof IExtensionInit initInstance) {
+            initInstance.close();
+            initInstance.init(naksha, extension);
+            instance = initInstance;
+          } else {
+            logger.error("Extension does not implement IExtensionInit for extension {}", extension.getId());
+            return;
+          }
+        } catch (Exception e) {
           logger.error(
               "Failed to instantiate class {} for extension {} ",
               extension.getInitClassName(),
-              extension.getId(),
+              extensionIdWthEnv,
               e);
           return;
         }
       }
       if (!isNullOrEmpty(extension.getInitClassName()))
         logger.info(
             "Extension {} initialization using initClassName {} done successfully.",
-            extension.getId(),
+            extensionIdWthEnv,
             extension.getInitClassName());
-      loaderCache.put(extension.getId(), new KVPair<Extension, ClassLoader>(extension, loader));
-      PluginCache.removeExtensionCache(extension.getId());
+
+      loaderCache.put(extensionIdWthEnv, new ValueTuple(extension, loader, instance));
+      PluginCache.removeExtensionCache(extensionIdWthEnv);
+
       logger.info(
           "Extension id={}, version={} is successfully loaded into the cache, using Jar at {} for env={}.",
-          extension.getId(),
+          extensionIdWthEnv,
           extension.getVersion(),
           extension.getUrl().substring(extension.getUrl().lastIndexOf("/") + 1),
-          extensionConfig.getEnv());
+          extension.getEnv());
+    }
+  }
+
+  private void removeExtensionFromCache(String extensionId) {
+    ValueTuple valueTuple = loaderCache.get(extensionId);
+    if (valueTuple != null) {
+      IExtensionInit instance = valueTuple.getInstance();
+      Extension extension = valueTuple.getExtension();
+      final String extensionIdWithEnv = extension.getEnv() + ":" + extension.getId();
```

Comment:
function parameter extensionId should already have env value. So, I think we don't need to concatenate env value again.

Reply:
Yes correct. Updated it.